ecl/hqlcpp/hqlhtcpp.cpp

```diff
@@ -14113,43 +14129,51 @@ ABoundActivity * HqlCppTranslator::doBuildActivityDedup(BuildCtx & ctx, IHqlExpr
         else
-            instance->nestedctx.addQuotedLiteral("virtual ICompare * queryKeyCompare() { return &Compare; }");
-
-        //virtual unsigned getFlags() = 0;
         {
-            StringBuffer flags;
-            if (recordTypesMatch(dataset, keyDataset)) flags.append("|HFDwholerecord");
-            if (flags.length())
-                doBuildUnsignedFunction(instance->classctx, "getFlags", flags.str()+1);
+            //virtual ICompare * queryKeyCompare()
+            OwnedHqlExpr keyOrder = createValueSafe(no_sortlist, makeSortListType(NULL), selects);
+            if (recordTypesMatch(dataset, keyDataset))
+            {
+                OwnedHqlExpr globalOrder = replaceSelector(order, dataset, queryActiveTableSelector());
+                if (keyOrder == globalOrder)
+                    reuseCompare = true;
+            }
+            if (!reuseCompare)
+            {
+                buildCompareMember(instance->nestedctx, "KeyCompare", keyOrder, DatasetReference(keyDataset, no_activetable, NULL));
+                buildHashOfExprsClass(instance->nestedctx, "KeyHash", keyOrder, DatasetReference(keyDataset, no_activetable, NULL), true);
+                //virtual ICompare * queryRowKeyCompare()=0; // lhs is a row, rhs is a key
+                doCompareLeftRight(instance->nestedctx, "RowKeyCompare", DatasetReference(dataset), DatasetReference(keyDataset, no_activetable, NULL), info.equalities, selects);
+            }
+            bHFDwholerecord = recordTypesMatch(dataset, keyDataset);
         }
 
-        //virtual IHash    * queryKeyHash()=0;
         if (reuseCompare)
+        {
+            instance->nestedctx.addQuotedLiteral("virtual ICompare * queryKeyCompare() { return &Compare; }");
             instance->nestedctx.addQuotedLiteral("virtual IHash * queryKeyHash() { return &Hash; }");
-        else
-            buildHashOfExprsClass(instance->nestedctx, "KeyHash", keyOrder, DatasetReference(keyDataset, no_activetable, NULL), true);
+            instance->nestedctx.addQuotedLiteral("virtual ICompare * queryRowKeyCompare() { return &Compare; }");
+        }
 
-        //virtual ICompare * queryRowKeyCompare()=0; // lhs is a row, rhs is a key
-        if (!reuseCompare)
         {
-            doCompareLeftRight(instance->nestedctx, "RowKeyCompare", DatasetReference(dataset), DatasetReference(keyDataset, no_activetable, NULL), info.equalities, selects);
+            StringBuffer flags;
+            if (bHFDwholerecord) flags.append("|HFDwholerecord");
+            if (info.keepBest) flags.append("|DFkeepbest");
+            if (flags.length())
+                doBuildUnsignedFunction(instance->classctx, "getFlags", flags.str()+1);
         }
-        else
-            instance->nestedctx.addQuotedLiteral("virtual ICompare * queryRowKeyCompare() { return &Compare; }");
+    }
 
+    if (info.keepBest)
+    {
+        OwnedHqlExpr sortOrder (LINK( expr->queryAttribute(bestAtom)->queryChild(0)));
```

Comment:
style:  Strange whitespace.  But better written as

    LinkedHqlExpr sortOrder(expr->queryAttribute(bestAtom)->queryChild(0));

but since sortOrder is never modified, even better as

    IHqlExpression * sortOrder = expr->queryAttribute(bestAtom)->queryChild(0);